Expose ConfigProperties to Scout JS

mvilliger opened the pull request:
- Add new config utility in Scout JS to access config properties from backend systems
- Add REST resource and ResourceLoader to fetch selected config properties from UI server and backend server
- Add contributor to expose config properties
- Add function to convert IConfigProperties to ConfigPropertyDo
- Replace existing properties sent to the browser with new config utility:
  - devMode
  - application.version
  - ui.backgroundPollingMaxWaitTime
  - uinotification.waitTimeout

373066

mvilliger committed:
Update browser cache on CodeType and Permission invalidate

373498

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/config/AbstractConfigPropertyToDoFunction.java:
```diff
@@ -0,0 +1,42 @@
+ * Base implementation of a function that converts {@link IConfigProperty} to {@link ConfigPropertyDo}. Only
+ * {@link IConfigProperty#getKey()} and {@link IConfigProperty#getValue()} is exported.
+ */
+public abstract class AbstractConfigPropertyToDoFunction<EXPLICIT_SOURCE extends IConfigProperty<?>, EXPLICIT_TARGET extends ConfigPropertyDo>
```
I recently moved some other ToDo-functions to a ".mapping" package to separate them from the business objects like permissions or config properties here. Suggestion: move these new functions to a .mapping package as well.

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/config/ConfigPropertyDo.java:
```diff
@@ -0,0 +1,75 @@
+   *     The property to convert.
+   * @return The corresponding {@link ConfigPropertyDo} or {@code null} if it cannot be converted.
+   */
+  public static ConfigPropertyDo create(IConfigProperty<?> property) {
```
As long as this method is just called in AbstractApiExposedConfigPropertyContributor.convert() I would move this helper code there and remove this method on this class.
Other API Data Objects which do not know the correspondent business object (like permission) also do not provide such convenience.

paolobazzi commented on org.eclipse.scout.rt.api.data/src/main/java/org/eclipse/scout/rt/api/data/config/ConfigPropertyDo.java:
```diff
@@ -0,0 +1,75 @@
+ * Do for {@link IConfigProperty}. Use {@link #create(IConfigProperty)} to convert a property to a DO.
+ */
+@TypeName("scout.ConfigProperty")
+public class ConfigPropertyDo extends DoEntity {
```
Add disclaimer that this DO should only be used to return config properties as response to a ScoutJS REST Client.
A Java REST layer will (without any further type information about the value) not be able to deserialize an arbitrary object value.
This also will just work for properties having a Jackson-serializable value type. An arbitrary POJO is not serializable as JSON.
As a hint for the caller, adding type information as additional field or using different config property DOs for various base types would be helpful.

cguglielmo commented on eclipse-scout-core/src/code/codes.ts:
```diff
@@ -29,6 +29,14 @@ export const codes = {
   },
 
+  /**
+   * load codes from the main system
```
Please use third person -> Loads. Same for bootstrap method
Also write it more clearly what it does, as you have done with config.ts

cguglielmo commented on eclipse-scout-core/src/config/config.ts:
```diff
@@ -0,0 +1,116 @@
+  },
+
+  /**
+   * Load config properties from the given system using the configuration of that {@link System}.
```
Third person, same for init

cguglielmo commented on eclipse-scout-core/src/config/config.ts:
```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Gets the value of the property with given key. Optionally from a specific backend system.
+   * <p>
```
Use markdown syntax (not html)

cguglielmo commented on eclipse-scout-core/src/config/config.ts:
```diff
@@ -0,0 +1,116 @@
+
+  /**
+   * Loads the properties from the system given and returns the property with given key.
+   * <p>
```
remove html

cguglielmo commented on eclipse-scout-core/src/config/config.ts:
```diff
@@ -0,0 +1,116 @@
+  },
+
+  /**
+   * Loads the properties from the system given and returns the property with given key.
```
system given -> given system
with given key -> with the given key

cguglielmo commented on eclipse-scout-core/src/config/config.ts:
```diff
@@ -0,0 +1,116 @@
+   * @param key The key of the property.
+   * @param system An optional system to which the property belongs. By default, {@link System.MAIN_SYSTEM} is used.
+   */
+  load<TKey extends keyof ConfigProperties[TSystem] & string, TSystem extends keyof ConfigProperties & string = 'main'>(key: TKey, system?: TSystem): JQuery.Promise<ConfigProperties[TSystem][TKey]> {
```
the first call writes the properties into the registry. If the property changes and is loaded a second time, there will be a warning, won't it? 
Wouldn't it be better to only load the desired property (not all) without writing it into the registry?

mvilliger replied:
The idea of this solution is that you have the possibility to preload the properties from a system, if you know that you won't need it before and afterwards you can access them sync. Therefore all properties are cached to allow a later sync access. The warning is changed to a info log in a subsequent commit. The old value is then overwritten.

cguglielmo commented on eclipse-scout-core/src/security/access.ts:
```diff
@@ -24,6 +24,11 @@ export const access = {
     return accessControl.whenSync();
   },
 
+  bootstrapSystem(): JQuery.Promise<void> {
```
Add similar jsdoc as for config.bootstrap

cguglielmo commented on eclipse-scout-core/src/system/System.ts:
```diff
@@ -0,0 +1,158 @@
+    return this._concatPath(this.baseUrl, relEndpointUrl);
+  }
+
+  protected _concatPath(a: string, b: string): string {
```
Maybe simplify using strings.removePrefix, strings.removeSuffix, array.join?

cguglielmo commented on eclipse-scout-core/src/system/systems.ts:
```diff
@@ -0,0 +1,41 @@
+ * A {@link System} represents a REST backend (typically Scout UI or server backend).
+ */
+export const systems = {
+  _systemsMap: new Map<string, System>(),
```
Move outside of object to make it private?

mvilliger replied:
It was added to the object as private map intentionally because it simplifies debugging (e.g. easy inspect of the map in the console)

cguglielmo commented on eclipse-scout-core/src/system/systems.ts:
```diff
@@ -0,0 +1,41 @@
+  /**
+   * Retrieves (and creates if not yet existing) the {@link System} with given name.
+   * @param name The optional name of the System. If omitted, {@link System.MAIN_SYSTEM} is used.
+   * @param baseUrl The base URL of the REST resources of the system. This is the (typically relative) URL under which all the REST endpoints are available. If omitted, 'api/' is used as a default.
```
'api/' -> 'api'. Otherwise, it looks like a slash is required when registering a system

cguglielmo commented on scout-hellojs-app/src/main/resources/archetype-resources/__rootArtifactId__.app/src/main/js/__simpleArtifactName__.ts:
```diff
@@ -1,9 +1,20 @@
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
```
remove copyright header

cguglielmo commented on eclipse-scout-core/src/App.ts:
```diff
@@ -385,10 +392,7 @@ export class App extends EventEmitter {
-      this.version,
-      options.version,
-      $('scout-version').data('value'));
+    this.version = scout.nvl(this.version, options.version, config.get('scout.application.version'));
```
I don't think the version tag in the html can be replaced -> we need to discuss it

mvilliger committed:
Expose ConfigProperties to Scout JS

- Add new config utility in Scout JS to access config properties from
  backend systems
- Add new systems utility to configure backend API urls
- Add REST resource and ResourceLoader to fetch exposed config
  properties from UI server and backend server
- Add contributor to expose config properties
- Add function to convert IConfigProperties to ConfigPropertyDo
- Replace existing properties sent to the browser with new config
  utility:
  - devMode
  - ui.backgroundPollingMaxWaitTime
  - uinotification.waitTimeout

373066
